File: archive/s/swift/convex-hull.swift
import Foundation

let usage = """
    Usage: please provide at least 3 x and y coordinates as separate lists (e.g. "100, 440, 210")
    """

struct Point: Comparable {
    let x: Int
    let y: Int

    static func < (lhs: Point, rhs: Point) -> Bool {
        return lhs.x < rhs.x || (lhs.x == rhs.x && lhs.y < rhs.y)
    }
}

enum Orientation {
    case clockwise, counterClockwise, collinear
}

func orientation(_ a: Point, _ b: Point, _ c: Point) -> Orientation {
    let value = (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x)

    if value == 0 { return .collinear }
    return value > 0 ? .counterClockwise : .clockwise
}

func convexHull(from points: [Point]) -> [Point] {
    guard points.count > 2 else { return points }

    let sorted = points.sorted()

    func buildHalfHull(_ points: some Sequence<Point>) -> [Point] {
        var hull: [Point] = []
        hull.reserveCapacity(points.underestimatedCount)

        for p in points {
            while hull.count >= 2
                && orientation(hull[hull.count - 2], hull[hull.count - 1], p)
                    != .counterClockwise
            {
                hull.removeLast()
            }
            hull.append(p)
        }

        return hull
    }

    let lower = buildHalfHull(sorted)
    let upper = buildHalfHull(sorted.reversed())

    return lower.dropLast() + upper.dropLast()
}

extension StringProtocol {
    var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
}

func parseList(_ input: String) -> [Int]? {
    let parts = input.split(separator: ",")
    let values = parts.map { Int($0.trimmed) }

    guard values.count >= 3,
        values.allSatisfy({ $0 != nil })
    else {
        return nil
    }

    return values.compactMap { $0 }
}

func parsePoints(from args: [String]) -> [Point]? {
    guard args.count == 3,
        let xs = parseList(args[1]),
        let ys = parseList(args[2]),
        xs.count == ys.count
    else { return nil }

    return zip(xs, ys).map(Point.init)
}

guard let points = parsePoints(from: CommandLine.arguments) else {
    print(usage)
    exit(1)
}

print(convexHull(from: points).map { "(\($0.x), \($0.y))" }.joined(separator: "\n"))

File: archive/p/perl/convex-hull.pl
#!/usr/bin/env perl
use v5.42;

use feature 'keyword_any';
no warnings 'experimental::keyword_any';

sub usage {
    say 'Usage: please provide at least 3 x and y coordinates as separate lists (e.g. "100, 440, 210")';
    exit 1;
}

sub parse_list ($s) {
    return undef unless defined $s;

    my @vals = split /\s*,\s*/, $s;

    return undef if @vals < 3;
    return undef if any { $_ !~ /\A-?\d+\z/ } @vals;

    return [ map 0 + $_, @vals ];
}

sub cross ( $o, $a, $b ) {
    ( $a->[0] - $o->[0] ) * ( $b->[1] - $o->[1] ) -
      ( $a->[1] - $o->[1] ) * ( $b->[0] - $o->[0] );
}

sub build_half ( $points, $reverse = false ) {
    my @hull;

    for my $p ( $reverse ? reverse @$points : @$points ) {
        pop @hull
          while ( @hull >= 2 && cross( $hull[-2], $hull[-1], $p ) <= 0 );
        push @hull, $p;
    }

    pop @hull;
    return @hull;
}

sub convex_hull ($points) {
    my @pts = sort { $a->[0] <=> $b->[0] || $a->[1] <=> $b->[1] } @$points;

    my @lower = build_half( \@pts );
    my @upper = build_half( \@pts, true );

    return [ @lower, @upper ];
}

my ( $xs, $ys ) = @ARGV;

my $x = parse_list($xs);
my $y = parse_list($ys);

usage() unless $x && $y && @$x == @$y;
usage() if @$x < 3;

my @points = map { [ $x->[$_], $y->[$_] ] } 0 .. $#$x;

my $hull = convex_hull( \@points );

say "($_->[0], $_->[1])" for @$hull;
